core/events.js

```diff
@@ -866,6 +868,146 @@ Blockly.Events.Move.prototype.run = function(forward) {
   }
 };
 
+/**
+ * Class for a block drag event. Fired when block dragged into or out of
+ * the blocks UI.
+ * @param {Blockly.Block} block The moved block.  Null for a blank event.
+ * @extends {Blockly.Events.Abstract}
+ * @constructor
+ */
+Blockly.Events.DragOutside = function(block) {
+  if (!block) {
+    return;  // Blank event to be populated by fromJson.
+  }
+  Blockly.Events.DragOutside.superClass_.constructor.call(this, block);
+};
+goog.inherits(Blockly.Events.DragOutside, Blockly.Events.Abstract);
+
+/**
+ * Class for a block drag event.
+ * @param {Blockly.Block} block The moved block.  Null for a blank event.
+ * @extends {Blockly.Events.Abstract}
+ * @constructor
+ */
+Blockly.Events.BlockDragOutside = Blockly.Events.DragOutside;
+
+/**
+ * Type of this event.
+ * @type {string}
+ */
+Blockly.Events.DragOutside.prototype.type = Blockly.Events.DRAG_OUTSIDE;
+
+/**
+ * Encode the event as JSON.
+ * @return {!Object} JSON representation.
+ */
+Blockly.Events.DragOutside.prototype.toJson = function() {
+  var json = Blockly.Events.DragOutside.superClass_.toJson.call(this);
+  if (this.isOutside) {
+    json['isOutside'] = this.isOutside;
+  }
+  return json;
+};
+
+/**
+ * Decode the JSON event.
+ * @param {!Object} json JSON representation.
+ */
+Blockly.Events.DragOutside.prototype.fromJson = function(json) {
+  Blockly.Events.DragOutside.superClass_.fromJson.call(this, json);
+  this.isOutside = json['isOutside'];
+};
+
+/**
+ * Does this event record any change of state?
+ * @return {boolean} True if something changed.
+ */
+Blockly.Events.DragOutside.prototype.isNull = function() {
+  return false;
+};
+
+/**
+ * Run a drag outside event.
+ */
+Blockly.Events.DragOutside.prototype.run = function() {
+  console.error('Not implemented');
```

Comment:
@rachel-fenichel when is this run function used? Should I implement it?

Reply:
Run is called to undo and redo events.  You don't need to implement it at all, because the version implemented in the base class is a no-op and that's what you want.